### DISCUSSION TITLE
Why Sometimes FCM Token is not getting generated?

### DISCUSSION THREAD

**Discussion by shasranj** [firebase/firebase-ios-sdk#11809] [Q&A]
Sometimes when we delete the app FCM token is not getting generating. Can you please help what might be the possible cause of it?

Steps Which we are following:
1. Method swizzling is Enabled

2. In didFinishLaunchingWithOptions methods we are calling:

FirebaseApp.configure()
Messaging.messaging().delegate = self

UNUserNotificationCenter.current().delegate = self

3. Delegated is implemented, but sometimes not getting FCM token in it:

 func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
         debugPrint("Firebase registration token: \(fcmToken)")
}

Also want to understand , what got changed in this version(Version 10.12.0 - July 11, 2023), FCM Token related.
<img width="1211" alt="Screenshot 2023-09-14 at 6 07 15 PM" src="https://github.com/firebase/firebase-ios-sdk/assets/145020548/80d07289-8397-498f-9e79-152fe9205396">

**Comment by morganchen12** [comment]
The change in the changelog does not affect FCM token generation. Are you able to receive an APNs token? https://developer.apple.com/documentation/usernotifications/registering_your_app_with_apns

**Reply by shasranj** [reply]
I am not receiving FCM token,
According to one of the thread in firebase i got this point, but not sure if it is 100% correct or not.

"Firebase Messaging automatically handles APNs token changes when swizzling is enabled, so you generally don't need to manually call registerForRemoteNotifications."
Link: https://github.com/firebase/firebase-ios-sdk/discussions/10685

**Reply by morganchen12** [reply]
The APNs token must still be generated by the operating system before it can be associated with an FCM token. Do you have an APNs token?

**Reply by shasranj** [reply]
Q1.So are you suggesting we have to call UIApplication.shared.registerForRemoteNotifications() in didFinishLaunchingWithOptions methods to get APNs token ?
Q2. Also need to know do we need to call UIApplication.shared.registerForRemoteNotifications() every time on didFinishLaunchingWithOptions methods or just once in the mobile app lifecycle. ?

**Reply by morganchen12** [reply]
You ~~should not~~ do need to manually call that method. If you do, nothing bad happens. Yes, it should be called when your app is launched and only once for the app's entire lifetime. Are you receiving anything in [this UIApplicationDelegate method](https://developer.apple.com/documentation/watchkit/wkapplicationdelegate/3946538-didregisterforremotenotification) if you implement it?

**Reply by shasranj** [reply]
Scenario 1:
1. Call UIApplication.shared.registerForRemoteNotifications()
2. didRegisterForRemoteNotificationsWithDeviceToken is getting called with apns token .

Scenario 2:
1. Didn't call UIApplication.shared.registerForRemoteNotifications()
2. didRegisterForRemoteNotificationsWithDeviceToken is not getting called .

So can you please suggest, to successfully generate FCM do we need to follow Scenario 1 Or Scenario 2 ?

**Reply by morganchen12** [reply]
Use scenario 1. Apologies, I mistakenly said you do not need to call `registerForRemoteNotifications`, which is wrong. Calling `registerForRemoteNotifications` is required.
